### PR TITLE
Adding the capability to run tests against a published dll

### PR DESCRIPTION
**Pull request by livarcocc**
Added abstractions to resolve the test runner based on different criteria. Before, we only resolved it from the project.json, now we look for a runner in the published directory of the app or use a parameter passed in.

Also added a new set of CommandResolvers that resolves a command based on the published directory of the app.

cc @brthor @eerhardt @piotrpMSFT @Sridhar-MS

### CODE REVIEW

File: src/dotnet-test-console/project.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      "target": "project"
+    },
+    "Microsoft.DotNet.InternalAbstractions": {
+  	  "target": "project"
+  	},
+    "Microsoft.DotNet.PlatformAbstractions": "1.0.1-beta-000919"
+  },
+
+  "frameworks": {
+    "netstandard1.6": {
```
Comment (eerhardt) [inline src/dotnet-test-console/project.json]: Is the point of splitting this into its own assembly that it is a standalone executable?

Basically, is "dotnet-test-console" an executable or a library?  If it is an executable, I think it should target 'netcoreapp1.0', have "emitEntryPoint: true" and have a "Main" method.

File: src/dotnet-test-console/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+
+[assembly: AssemblyTitle("dotnet_test_console")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
+[assembly: AssemblyProduct("$safeprojectname$")]
```
Comment (eerhardt) [inline src/dotnet-test-console/Properties/AssemblyInfo.cs]: I don't think a lot of these attributes make sense and should be removed.

The only ones you need are 'InternalsVisibleTo' and [AssemblyMetadata('Serviceable', 'True')] (which is missing).

File: src/Microsoft.Extensions.Testing.Abstractions/Properties/AssemblyInfo.cs [hunk 3]
```diff
@@ -5,4 +5,4 @@
 using System.Resources;
 
 [assembly: AssemblyMetadata("Serviceable", "True")]
-[assembly: NeutralResourcesLanguage("en-us")]
\ No newline at end of file
+[assembly: NeutralResourcesLanguage("en-us")]
```
Comment (eerhardt) [inline src/Microsoft.Extensions.Testing.Abstractions/Properties/AssemblyInfo.cs]: unnecessary change in this file

File: src/dotnet-test-console/project.json (hunk 1 above)
Comment (eerhardt) [inline src/dotnet-test-console/project.json]: If it is a library, I think it should be named like a library.

---
In reply to: [76426587](https://github.com/dotnet/cli/pull/4102#discussion_r76426587) [](ancestors = 76426587)

Comment (livarcocc) [inline src/dotnet-test-console/project.json]: You are right. I will rename it. I started with an executable but then changed to a library, but didn't update the naming. Will do.

File: src/dotnet-test-console/Properties/AssemblyInfo.cs (hunk 2 above)
Comment (livarcocc) [inline src/dotnet-test-console/Properties/AssemblyInfo.cs]: Agreed. These is what ProjectRyder created. I will clean it up.

File: src/Microsoft.Extensions.Testing.Abstractions/Properties/AssemblyInfo.cs (hunk 3 above)
Comment (livarcocc) [inline src/Microsoft.Extensions.Testing.Abstractions/Properties/AssemblyInfo.cs]: yeah, not sure why this happened.

File: Microsoft.DotNet.Cli.sln [hunk 4]
```diff
@@ -156,6 +152,8 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "publish", "publish", "{27B1
 		build\publish\PublishContent.targets = build\publish\PublishContent.targets
 	EndProjectSection
 EndProject
+Project("{8BB2217D-0F2D-49D1-97BC-3654ED321F3B}") = "dotnet-test-console", "src\dotnet-test-console\dotnet-test-console.xproj", "{6D028154-5518-4A56-BAD6-938A90E5BCF6}"
```
Comment (eerhardt) [inline Microsoft.DotNet.Cli.sln]: this is out of date.

File: src/Microsoft.DotNet.Cli.Utils/CommandResolution/PublishedPathCommandResolver.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.IO;
+
+namespace Microsoft.DotNet.Cli.Utils
+{
+    public class PublishedPathCommandResolver : ICommandResolver
```
Comment (eerhardt) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/PublishedPathCommandResolver.cs]: This doesn't actually have to be a "published" app right?  This would work against just a "built" app.

I'm wondering if "Published" is a confusing word here.  What exactly is the intention of the class?

Comment (eerhardt) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/PublishedPathCommandResolver.cs]: Maybe "ApplicationCommandResolver"??  "DotNetAppCommandResolver"??  

This can run any .NET Core app, right?

I might be reaching here...

---
In reply to: [76478100](https://github.com/dotnet/cli/pull/4102#discussion_r76478100) [](ancestors = 76478100)

File: test/dotnet-test.Tests/GivenThatWeWantToRunTestsInTheConsole.cs [hunk 6]
```diff
@@ -97,6 +97,20 @@ public void It_skips_build_when_the_no_build_flag_is_passed()
...
+        {
+            var publishCommand = new PublishCommand(_projectFilePath);
+            var result = publishCommand.Execute();
+            result.Should().Pass();
+
+            var assemblyUnderTestPath = Path.Combine(publishCommand.GetOutputDirectory(true).FullName, publishCommand.GetPortableOutputName());
+
+            var testCommand = new DotnetTestCommand();
+            result = testCommand.Execute($"{assemblyUnderTestPath}");
+            result.Should().Pass();
```
Comment (eerhardt) [inline test/dotnet-test.Tests/GivenThatWeWantToRunTestsInTheConsole.cs]: Is ensuring a non-error exit code enough to verify that the tests were run?

**Comment by eerhardt** [comment]
:shipit:

File: .gitignore [hunk 7]
```diff
@@ -151,6 +151,7 @@ _ReSharper*/
 *.DotSettings.user
 .idea/
 *.iml
+*.DotSettings
```
Comment (brthor) [inline .gitignore]: 👍

File: src/Microsoft.DotNet.Cli.Utils/CommandResolution/PublishPathCommandSpecFactory.cs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+            {
+                return CreatePackageCommandSpecUsingMuxer(
+                    commandPath,
+                    commandArguments,
+                    depsFilePath,
+                    commandResolutionStrategy,
+                    runtimeConfigPath);
+            }
+
+            return CreateCommandSpec(commandPath, commandArguments, commandResolutionStrategy);
```
Comment (brthor) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/PublishPathCommandSpecFactory.cs]: Is there a case where we'll be running arbitrary executables from the publish directory? desktop?

File: src/Microsoft.DotNet.ProjectModel/Properties/AssemblyInfo.cs [hunk 9]
```diff
@@ -27,3 +27,7 @@
 
 [assembly: AssemblyMetadataAttribute("Serviceable", "True")]
 [assembly: InternalsVisibleTo("Microsoft.DotNet.ProjectModel.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100039ac461fa5c82c7dd2557400c4fd4e9dcdf7ac47e3d572548c04cd4673e004916610f4ea5cbf86f2b1ca1cb824f2a7b3976afecfcf4eb72d9a899aa6786effa10c30399e6580ed848231fec48374e41b3acf8811931343fc2f73acf72dae745adbcb7063cc4b50550618383202875223fc75401351cd89c44bf9b50e7fa3796")]
+[assembly:
+  InternalsVisibleTo(
+    "dotnet-test.UnitTests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100039ac461fa5c82c7dd2557400c4fd4e9dcdf7ac47e3d572548c04cd4673e004916610f4ea5cbf86f2b1ca1cb824f2a7b3976afecfcf4eb72d9a899aa6786effa10c30399e6580ed848231fec48374e41b3acf8811931343fc2f73acf72dae745adbcb7063cc4b50550618383202875223fc75401351cd89c44bf9b50e7fa3796"
```
Comment (brthor) [inline src/Microsoft.DotNet.ProjectModel/Properties/AssemblyInfo.cs]: This may be a moot point but why do we strong name sign the unit tests?

File: src/Microsoft.DotNet.Cli.Utils/CommandResolution/PublishedPathCommandResolver.cs (hunk 5 above)
Comment (livarcocc) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/PublishedPathCommandResolver.cs]: It won't work for a built app, because it does not binplace your package dependencies. In which case, we would not be able to find commands, like dotnet-test-xunit. They only appear when you publish the app. Does that make sense?

File: src/Microsoft.DotNet.Tools.Test/TestRunners/AssemblyTestRunnerResolver.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+        internal AssemblyTestRunnerResolver(string assemblyUnderTest, IDirectory directory)
+        {
+            _directoryOfAssemblyUnderTest = directory.GetDirectoryFullName(assemblyUnderTest);
+            _directory = directory;
+        }
+
+        public string ResolveTestRunner()
+        {
+            var testRunnerPath = _directory.GetFiles(_directoryOfAssemblyUnderTest, "dotnet-test-*").FirstOrDefault();
```
Comment (brthor) [inline src/Microsoft.DotNet.Tools.Test/TestRunners/AssemblyTestRunnerResolver.cs]: If this returns default, you may get a null ref on the next call to Path.GetFileNameWithoutExtension

File: src/Microsoft.DotNet.Tools.Test/TestRunners/AssemblyTestRunnerResolver.cs [hunk 11]
```diff
@@ -0,0 +1,34 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.IO;
+using System.Linq;
+using Microsoft.Extensions.EnvironmentAbstractions;
+
+namespace Microsoft.DotNet.Tools.Test
+{
+    public class AssemblyTestRunnerResolver : ITestRunnerResolver
```
Comment (brthor) [inline src/Microsoft.DotNet.Tools.Test/TestRunners/AssemblyTestRunnerResolver.cs]: nit: I had slight confusion that this class was a `RunnerResolver` but yet only returned the name of the runner. What do you think of a slight name change to `AssemblyTestRunnerNameResolver` ?

File: src/Microsoft.DotNet.Tools.Test/ConsoleTestRunner.cs [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+            _testRunnerResolver = testRunnerResolver;
+            _commandFactory = commandFactory;
+            _assemblyUnderTest = assemblyUnderTest;
+            _framework = framework;
+        }
+
+        public int RunTests(DotnetTestParams dotnetTestParams)
+        {
+            return _commandFactory.Create(
+                    _testRunnerResolver.ResolveTestRunner(),
```
Comment (brthor) [inline src/Microsoft.DotNet.Tools.Test/ConsoleTestRunner.cs]: This can return null if the _testRunnerResolver is a `ProjectJsonTestRunnerResolver` and no _testRunner is specified, is this going to null ref?

File: src/Microsoft.DotNet.ProjectModel/Properties/AssemblyInfo.cs (hunk 9 above)
Comment (livarcocc) [inline src/Microsoft.DotNet.ProjectModel/Properties/AssemblyInfo.cs]: So that they can have internalvisibleto strong named assemblies.

Comment (brthor) [inline src/Microsoft.DotNet.ProjectModel/Properties/AssemblyInfo.cs]: Is that a desirable thing?

File: src/Microsoft.DotNet.Cli.Utils/CommandResolution/PublishPathCommandSpecFactory.cs (hunk 8 above)
Comment (brthor) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/PublishPathCommandSpecFactory.cs]: Agreed, on adding a test for this if we're enabling it

File: src/Microsoft.DotNet.ProjectModel/Properties/AssemblyInfo.cs (hunk 9 above)
Comment (livarcocc) [inline src/Microsoft.DotNet.ProjectModel/Properties/AssemblyInfo.cs]: Of course, how else can this test mock the classes inside InternalAbstractions. InternalAbstractions is signed and for it to make its internals visible to a test project, the test project also needs to be signed, which is why we have it signed.

File: src/Microsoft.DotNet.Tools.Test/TestRunners/AssemblyTestRunnerResolver.cs (hunk 10 above)
Comment (livarcocc) [inline src/Microsoft.DotNet.Tools.Test/TestRunners/AssemblyTestRunnerResolver.cs]: No, Path.GetFileNameWithoutExtension(null) returns null.

File: src/Microsoft.DotNet.Tools.Test/ConsoleTestRunner.cs (hunk 12 above)
Comment (livarcocc) [inline src/Microsoft.DotNet.Tools.Test/ConsoleTestRunner.cs]: No. It will fail to find a matching command of the format dotnet-test-. Because the projectjsontestrunnernameresolver will do a string format when it returns.

File: src/Microsoft.DotNet.Tools.Test/TestRunners/AssemblyTestRunnerResolver.cs (hunk 11 above)
Comment (livarcocc) [inline src/Microsoft.DotNet.Tools.Test/TestRunners/AssemblyTestRunnerResolver.cs]: Changed name here and for all other TestRunnerResolvers.

File: src/Microsoft.DotNet.Cli.Utils/CommandResolution/PublishedPathCommandResolver.cs (hunk 5 above)
Comment (eerhardt) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/PublishedPathCommandResolver.cs]: So then how do we expect "dotnet test" to work?  Does it need to publish your test project first?

I would think "dotnet test" only needs to "build" the project, and it should work.

**Comment by Yantrio** [comment]
Is there any way I can get ahold of this build? I would love to be able to run tests against a dll :(
